### PR TITLE
[GLT-3731] added text input filtering controls

### PR DESCRIPTION
**Pull request by Arcslogger**

### CODE REVIEW

**Review by mlaszloffy — APPROVED** [review]

File: ts/util/table-builder.ts [hunk 1]
```diff
@@ -255,6 +256,11 @@ export class TableBuilder<ParentType, ChildType> {
                 reload();
               })
           );
+          if (!filterSuboptions) {
+            throw new Error(
+              `Dropdown filter ${filter.key} has no menu options`
+            );
+          }
```
Comment (djcooke) [inline ts/util/table-builder.ts:263]: Move this check above the `const filterSuboptions...` line and change to `if (!filter.values || !filter.values.length)`. Then you don't need the `?` in `filter.values.map`, and it'll also protect against `filter.values` being an empty array

File: ts/util/text-input.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        textInputClickout.classList.toggle("hidden");
+        onClose(this);
+      } else {
+        window.setTimeout(() => this.textField.focus(), 0);
+      }
+    };
+    submitIcon.onclick = submitTextInput;
+    textInputClickout.onclick = submitTextInput;
+    this.textField.addEventListener("keypress", (event) => {
+      console.log(event.key);
```
Comment (djcooke) [inline ts/util/text-input.ts:37]: Remove

File: ts/util/text-input.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      "relative",
+      "z-10"
+    );
+
+    const submitTextInput = () => {
+      if (this.textField.value) {
+        textInputClickout.classList.toggle("hidden");
+        onClose(this);
+      } else {
+        window.setTimeout(() => this.textField.focus(), 0);
```
Comment (djcooke) [inline ts/util/text-input.ts:31]: I don't think `setTimeout` is required here?

File: ts/util/table-builder.ts [hunk 4]
```diff
@@ -271,7 +277,31 @@ export class TableBuilder<ParentType, ChildType> {
           );
           break;
         case "text":
```
Comment (djcooke) [inline ts/util/table-builder.ts:279]: `addFilterControls` has gotten pretty long. Maybe separate the cases out into their own functions? e.g.

```
function addFilterControls(...) {
  ...
  switch (filterType) {
    case "dropdown":
      addDropdownFilter(...);
      return;
    case "text":
      addTextFilter(...);
      return;
    default:
      ...
  }
}
```

**Review by djcooke — APPROVED** [review]

**Pull request merged**